StaticAnalyzer/views/ios/rules/swift_rules.py

```diff
@@ -375,4 +375,74 @@
         'owasp': '',
         'owasp-mstg': OWASP_MSTG['platform-5'],
     },
+    {
+        'desc': ('TLS 1.3 should be used. '
+                 'Detected old version - TLS 1.0.'),
+        'type': MatchType.string,
+        'string1': ('.TLSMinimumSupportedProtocolVersion = '
+                    'tls_protocol_version_t.TLSv10'),
+        'level': Level.high,
+        'match': Match.single_string,
+        'input_case': InputCase.exact,
+        'cvss': 4.3,
+        'cwe': CWE['CWE-310'],
+        'owasp': OWASP['m3'],
+        'owasp-mstg': OWASP_MSTG['network-2'],
+    },
+    {
+        'desc': ('TLS 1.3 should be used. '
+                 'Detected old version - TLS 1.1.'),
+        'type': MatchType.string,
+        'string1': ('.TLSMinimumSupportedProtocolVersion = '
+                    'tls_protocol_version_t.TLSv11'),
+        'level': Level.high,
+        'match': Match.single_string,
+        'input_case': InputCase.exact,
+        'cvss': 4.3,
+        'cwe': CWE['CWE-310'],
+        'owasp': OWASP['m3'],
+        'owasp-mstg': OWASP_MSTG['network-2'],
+    },
+    {
+        'desc': ('TLS 1.3 should be used. '
+                 'Detected old version - TLS 1.2.'),
+        'type': MatchType.string,
+        'string1': ('.TLSMinimumSupportedProtocolVersion = '
+                    'tls_protocol_version_t.TLSv12'),
+        'level': Level.warning,
+        'match': Match.single_string,
+        'input_case': InputCase.exact,
+        'cvss': 0.0,
+        'cwe': '',
+        'owasp': OWASP['m3'],
+        'owasp-mstg': OWASP_MSTG['network-2'],
+    },
+    {
+        'desc': ('DTLS 1.2 should be used. '
+                 'Detected old version - DTLS 1.0.'),
+        'type': MatchType.string,
+        'string1': ('.TLSMinimumSupportedProtocolVersion = '
+                    'tls_protocol_version_t.DTLSv10'),
+        'level': Level.high,
+        'match': Match.single_string,
+        'input_case': InputCase.exact,
+        'cvss': 4.3,
+        'cwe': CWE['CWE-310'],
+        'owasp': OWASP['m3'],
+        'owasp-mstg': OWASP_MSTG['network-2'],
+    },
+    {
+        'desc': ('Use of deprecated property tlsMinimumSupportedProtocol. '
+                 'To avoid potential security risks, use '
+                 'tlsMinimumSupportedProtocolVersion'),
+        'type': MatchType.string,
+        'string1': '.tlsMinimumSupportedProtocol =',
+        'level': Level.info,
+        'match': Match.single_string,
+        'input_case': InputCase.exact,
+        'cvss': 0.0,
+        'cwe': '',
+        'owasp': OWASP['m3'],
+        'owasp-mstg': OWASP_MSTG['network-2'],
+    },
```

Comment:
I'm just wondering if we could merge all these rules into one rule like:
```
{
        'desc': ('App may use old network encryption',
                     ' network encryption should always be in the newest version.'),
        'type': MatchType.string,
        'string1': '.tlsMinimumSupportedProtocol =',
        'string2': '.TLSMinimumSupportedProtocolVersion =',
        'level': Level.warning,
        'match': Match.string_or,
        'input_case': InputCase.exact,
        'cvss': 4.3,
        'cwe': CWE['CWE-310'],
        'owasp': OWASP['m3'],
        'owasp-mstg': OWASP_MSTG['network-2'],
},
```
What do you think?
@mtbrzeski  @ajinabraham

Reply:
Would be nice to merge, but there are a couple of issues.

- **warning** is always used, while in the original code there can be either info, warning, or high depending on the TLS/DTLS version. Same with CVSS score
- you can have a false-positive here (which assigns CVSS) if somebody assigns TLS 1.3 manually

Maybe we could merge the rule for TLS 1.0 and TLS 1.1 as one? And keep the rest for precision. Also depends on the input from @ajinabraham regarding _levels_ and assigning _cvss_ to each case.

Reply:
@mtbrzeski I like the idea of merging TLS1.0 and 1.1. Since they both share the level high and TLS 1.2 is of level medium.
Can we change the match to
`string_and_or` for TLS1.0,1.1
and
`string_and` for TLS1.2 and DTLS 1.2

Let's avoid ambiguity from `=` and whitespace
Something like
`string1`: `TLSMinimumSupportedProtocolVersion`
`string2`: `tls_protocol_version_t.TLSv12`